### PR TITLE
features: add feature test methods

### PR DESCRIPTION
**Pull request by zyga**
features: add module for querying enabled snapd features
    
This builds upon the ability to export enabled experimental features to
userspace. We have the Go side for controlling features via "snap set
core experimental.foo 1" and the C side for querying them but we still
need the Go counterpart. This patch adds one.

The API is extremely simple, consisting of a new type, SnapdFeature,
along with a IsEnabled method that performs a stat call.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

File: features/features.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	Layouts SnapdFeature = iota
+	// ParallelInstances controls availability installing a snap multiple times.
+	ParallelInstances
+	// Hotplug controls availability of dynamically creating slots based on system hardware.
+	Hotplug
+	// SnapdSnap controls possibility of installing the snapd snap.
+	SnapdSnap
+	// PerUserMountNamespace controls the persistence of per-user mount namespaces.
+	PerUserMountNamespace
+	// PerformanceMeasurements controls the collection of performance data.
```
Comment (mvo5) [inline features/features.go]: We don't have this yet, I think we should not add it yet.

File: features/features.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	PerformanceMeasurements
+)
+
+// String returns the name of a snapd feature.
+func (f SnapdFeature) String() string {
+	// The constants here must be synchronized with cmd/libsnap-confine-private/feature.c
```
Comment (mvo5) [inline features/features.go]: And they also need to be synced with overlord/configstate/configcore/experimental.go  - this makes me a bit uneasy. I wonder if there is a way to sync configcore and this closer. Right now we when adding a experimental flag we need to add things in three places AFAICT, I wonder if there is a way to share the GO bits more (and maybe even the C bits but that is of course way harder).

File: features/features.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		return "per-user-mount-namespace"
+	case PerformanceMeasurements:
+		return "performance-measurements"
+	}
+	panic(fmt.Sprintf("unknown feature flag code %d", f))
+}
+
+// IsEnabled checks if a given snapd feature is enabled.
+func (f SnapdFeature) IsEnabled() bool {
+	fi, err := os.Stat(filepath.Join(dirs.FeaturesDir, f.String()))
```
Comment (mvo5) [inline features/features.go]: Why not just `osutil.FileExists()`?

File: features/features.go (hunk 2 above)
Comment (zyga) [reply features/features.go]: As I said in the other PR I will follow up with using those constants there. We will only need to do two places (C and Go) which is IMO good enough for now.

File: features/features.go (hunk 3 above)
Comment (zyga) [reply features/features.go]: Ah, I thought we had something like that but couldn't find it quickly. I'll use it :)

File: features/features.go (hunk 1 above)
Comment (zyga) [reply features/features.go]: Sure

File: features/features.go (hunk 2 above)
Comment (pedronis) [reply features/features.go]: we are growing a bit too many small packages I fear.  If we have this the code in configcore/experimental.go should probably become dynamic and use listing here that should flag a feature as experimental and whether it's turned to default.

**Comment by zyga** [comment]
@pedronis this module needs to be imported from things outside of snapd. The changes to use it in configcore/experimenta.go are coming in another PR (6170 after this one lands).

**Comment by pedronis** [comment]
no, changes to experimental.go needs to happen together with this here

**Comment by pedronis** [comment]
my point is that if we have this, just using the constant is not enough, we need to have the full set of facts around features/experimental features here, and other places use the facts. I shouldn't need to go over there to know if a feature is experimental and whether is now default or not

**Comment by zyga** [comment]
Closing to integrate with https://github.com/snapcore/snapd/pull/6170

**Comment by zyga** [comment]
This is now proposed, unified, as https://github.com/snapcore/snapd/pull/6190

**Pull request closed without merging**